compiler/rustc_typeck/src/astconv/generics.rs

```diff
@@ -20,29 +20,27 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
     /// Report an error that a generic argument did not match the generic parameter that was
     /// expected.
     fn generic_arg_mismatch_err(
-        sess: &Session,
+        tcx: TyCtxt<'_>,
         arg: &GenericArg<'_>,
-        kind: &'static str,
+        param: &GenericParamDef,
         possible_ordering_error: bool,
         help: Option<&str>,
     ) {
+        let sess = tcx.sess;
         let mut err = struct_span_err!(
             sess,
             arg.span(),
             E0747,
             "{} provided when a {} was expected",
             arg.descr(),
-            kind,
+            param.kind.descr(),
         );
 
         let unordered = sess.features_untracked().const_generics;
-        let kind_ord = match kind {
-            "lifetime" => ParamKindOrd::Lifetime,
-            "type" => ParamKindOrd::Type,
-            "constant" => ParamKindOrd::Const { unordered },
-            // It's more concise to match on the string representation, though it means
-            // the match is non-exhaustive.
-            _ => bug!("invalid generic parameter kind {}", kind),
+        let kind_ord = match param.kind {
+            GenericParamDefKind::Lifetime => ParamKindOrd::Lifetime,
+            GenericParamDefKind::Type { .. } => ParamKindOrd::Type,
+            GenericParamDefKind::Const => ParamKindOrd::Const { unordered },
         };
 
         if let ParamKindOrd::Const { .. } = kind_ord {
```

Comment:
can't we just match on `param.kind` directly here :laughing:

Reply:
On line 46 yes, but I left the ord in cause we compare it with the arg ord later

Reply:
hmm

can you move both `GenericParamDefKind -> ParamKindOrd` and `GenericArg -> ParamKindOrd` to inherent methods (something like `fn to_param_kind_ord(&self, tcx: TyCtxt<'_>) -> ParamKindOrd`) for these types instead then? Thanks for your patience :bow: